ext/pcntl/pcntl.c

```diff
@@ -1365,28 +1363,25 @@ static void pcntl_signal_handler(int signo)
 				pid = waitpid(WAIT_ANY, &status, WNOHANG | WUNTRACED);
```

Comment:
@nielsdos Do you know if this can lead to data races? If two threads have the `SIGCHLD` signal handler installed, they might both be invoking `waitpid` at the same time, and the faster one might consume the children from the other thread. I'm not sure in how signals are handled with ZTS.

Reply:
@dunglas might know more about how signals are handled in ZTS as IIRC he did some work around signals recently.

Reply:
Signals are per process, so a signal is delivered to whatever thread, this is arbitrary as defined by POSIX. So the sample of having multiple threads is already broken probably even without my patch. So the signal will only be delivered to one thread, and there is no race for the *same* signal.
What happens with the waitpid is indeed that the arbitrary thread will consume all children, which is consistent with the already-broken behaviour afaik.

Reply:
@nielsdos What I meant is that if two children exit, and the two signals are dispatched to separate threads, two threads might be competing for `waitpid`.

Reply:
@iluuu1994 
Yes then they have a race in waitpid and whoever goes first wins. Which is an arbitrary thread.

However, if you revert the waitpid patch, the threads that receive the signals are still chosen arbitrarily. So this is not a new problem. 

I don't know how to solve that, threads and signals are kinda incompatible by design.

Reply:
Thank you for the explanation! I suspected that, but I thought it would be better to double check :slightly_smiling_face:

Reply:
In that case though, we should keep the `if (psig == psig_first) {` branch and handle it gracefully by suppressing the signal, as the other thread would receive it.

Reply:
Now we had a irl race condition in commenting ;)

Reply:
Haha :smile: Done, thanks for your feedback! I also added an echo to the handler just to prove that we have no superfluous calls.

Reply:
> a signal is delivered to whatever thread

Technically, it's possible to send a signal to a specific thread using (at least on Unix) using [`pthread_kill()`](https://www.man7.org/linux/man-pages/man3/pthread_kill.3.html), but it's probably not relevant here.

Reply:
@dunglas Good to know, thank you! In this case, the signal `SIGCHLD is sent by the kernel so this doesn't seem relevant. The adjustment discussed here would actually also be necessary for signals sent manually which I forgot about, as there would be no accompanying process to `waitpid`.